pandas/core/exchange/dataframe.py

```diff
@@ -0,0 +1,101 @@
+from collections import abc
+
+import pandas as pd
+from pandas.core.exchange.column import PandasColumn
+from pandas.core.exchange.dataframe_protocol import DataFrame as DataFrameXchg
+
+
+class PandasDataFrameXchg(DataFrameXchg):
+    """
+    A data frame class, with only the methods required by the interchange
+    protocol defined.
+    Instances of this (private) class are returned from
+    ``pd.DataFrame.__dataframe__`` as objects with the methods and
+    attributes defined on this class.
+    """
+
+    def __init__(
+        self, df: pd.DataFrame, nan_as_null: bool = False, allow_copy: bool = True
+    ) -> None:
+        """
+        Constructor - an instance of this (private) class is returned from
+        `pd.DataFrame.__dataframe__`.
+        """
+        self._df = df
+        # ``nan_as_null`` is a keyword intended for the consumer to tell the
+        # producer to overwrite null values in the data with ``NaN`` (or ``NaT``).
+        # This currently has no effect; once support for nullable extension
+        # dtypes is added, this value should be propagated to columns.
+        self._nan_as_null = nan_as_null
+        self._allow_copy = allow_copy
+
+    def __dataframe__(self, nan_as_null: bool = False, allow_copy: bool = True):
+        return PandasDataFrameXchg(self._df, nan_as_null, allow_copy)
+
+    @property
+    def metadata(self):
+        # `index` isn't a regular column, and the protocol doesn't support row
+        # labels - so we export it as Pandas-specific metadata here.
+        return {"pandas.index": self._df.index}
+
+    def num_columns(self) -> int:
+        return len(self._df.columns)
+
+    def num_rows(self) -> int:
+        return len(self._df)
+
+    def num_chunks(self) -> int:
+        return 1
+
+    def column_names(self):
+        return self._df.columns.tolist()
```

Comment:
why `.tolist()` is needed? If it's protection from a user modifying the metadata of underlying frame, then maybe it's enough to do an explicit copy?
```python
    return self._df.columns.copy()
```

Reply:
I've looked up and I think it should be enough to just drop `.tolist()`, as `Index` should already be immutable.